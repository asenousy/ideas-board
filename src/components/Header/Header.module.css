.wrapper {
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
  align-items: center;
  background-image: linear-gradient(103deg, #6d29c5, #d54ce0);
  border-radius: 10px;
  margin: 10px;
  padding: 10px;
  color: white;
}

.label {
  margin: 0 10px;
  font-size: 20px;
}

.select {
  margin: 0 10px;
}
