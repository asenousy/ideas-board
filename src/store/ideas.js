import { createSlice } from "redux-starter-kit";
import { sanitize } from "dompurify";

const todoSlice = createSlice({
  initialState: [],
  reducers: {
    addIdea: (state, { payload }) => [...state, payload],
    updateIdea: (state, { payload }) =>
      state.map(idea =>
        idea.id === payload.id
          ? {
              id: idea.id,
              // sanitizing user input for xss,
              // even though react render does sanitize it,
              //but in case this data will be passed somewhere in future
              title: sanitize(payload.title),
              description: sanitize(payload.description),
              date: payload.date,
              new: false
            }
          : idea
      ),
    deleteIdea: (state, { payload }) =>
      state.filter(idea => idea.id !== payload.id)
  }
});

export default todoSlice;
