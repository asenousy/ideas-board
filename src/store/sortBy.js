import { createSlice } from "redux-starter-kit";

export const sortEnum = { date: "date", alphabetical: "alphabetical" };

const sortBy = createSlice({
  initialState: sortEnum.date,
  reducers: {
    setSortBy: (state, { payload }) => payload
  }
});

export default sortBy;
