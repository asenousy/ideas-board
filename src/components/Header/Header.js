import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import styles from "./Header.module.css";
import { actions, sortEnum } from "../../store";
const { setSortBy } = actions;

export class Header extends PureComponent {
  render() {
    const { sortBy } = this.props;

    return (
      <div className={styles.wrapper}>
        <h1 className={styles.label}>Ideas Board:</h1>
        <label className={styles.label}>
          Sort by:
          <select
            className={styles.select}
            value={sortBy}
            onChange={this.handleChange}
          >
            <option value={sortEnum.date}>{sortEnum.date}</option>
            <option value={sortEnum.alphabetical}>
              {sortEnum.alphabetical}
            </option>
          </select>
        </label>
      </div>
    );
  }

  handleChange = e => {
    this.props.setSortBy && this.props.setSortBy(e.target.value);
  };
}

Header.propTypes = {
  sortBy: PropTypes.oneOf([sortEnum.date, sortEnum.alphabetical]).isRequired,
  setSortBy: PropTypes.func
};

export default connect(
  state => ({ sortBy: state.sortBy }),
  { setSortBy }
)(Header);
