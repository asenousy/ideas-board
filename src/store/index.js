import { combineReducers } from "redux";
import ideas from "./ideas";
import sortBy from "./sortBy";

export const reducer = combineReducers({
  ideas: ideas.reducer,
  sortBy: sortBy.reducer
});

export const actions = {
  ...ideas.actions,
  ...sortBy.actions
};

export { sortEnum } from "./sortBy";
