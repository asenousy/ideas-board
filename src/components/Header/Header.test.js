import React from "react";
import { shallow } from "enzyme";
import { Header } from "./Header";
import { sortEnum } from "../../store";

it("renders without crashing", () => {
  const wrapper = shallow(<Header sortBy={sortEnum.date} />);
  expect(wrapper).toMatchSnapshot();

  wrapper.setProps({ sortBy: sortEnum.alphabet });
  expect(wrapper).toMatchSnapshot();
});

it("updates sort state when select changes", () => {
  const mockFn = jest.fn();
  const wrapper = shallow(<Header sortBy={sortEnum.date} setSortBy={mockFn} />);
  wrapper
    .find("select")
    .simulate("change", { target: { value: sortEnum.date } });

  expect(mockFn).toHaveBeenCalledWith(sortEnum.date);
});
