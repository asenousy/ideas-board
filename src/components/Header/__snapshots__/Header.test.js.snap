// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Header
    sortBy="Date"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1
          className="title"
        >
          Ideas Board:
        </h1>,
        <label
          className="label"
        >
          Sort by:
          <select
            className="select"
            onChange={[Function]}
            value="Date"
          >
            <option
              value="Date"
            >
              Date
            </option>
            <option
              value="Alphabet"
            >
              Alphabet
            </option>
          </select>
        </label>,
      ],
      "className": "wrapper",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Ideas Board:",
          "className": "title",
        },
        "ref": null,
        "rendered": "Ideas Board:",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Sort by:",
            <select
              className="select"
              onChange={[Function]}
              value="Date"
            >
              <option
                value="Date"
              >
                Date
              </option>
              <option
                value="Alphabet"
              >
                Alphabet
              </option>
            </select>,
          ],
          "className": "label",
        },
        "ref": null,
        "rendered": Array [
          "Sort by:",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <option
                  value="Date"
                >
                  Date
                </option>,
                <option
                  value="Alphabet"
                >
                  Alphabet
                </option>,
              ],
              "className": "select",
              "onChange": [Function],
              "value": "Date",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Date",
                  "value": "Date",
                },
                "ref": null,
                "rendered": "Date",
                "type": "option",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Alphabet",
                  "value": "Alphabet",
                },
                "ref": null,
                "rendered": "Alphabet",
                "type": "option",
              },
            ],
            "type": "select",
          },
        ],
        "type": "label",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1
            className="title"
          >
            Ideas Board:
          </h1>,
          <label
            className="label"
          >
            Sort by:
            <select
              className="select"
              onChange={[Function]}
              value="Date"
            >
              <option
                value="Date"
              >
                Date
              </option>
              <option
                value="Alphabet"
              >
                Alphabet
              </option>
            </select>
          </label>,
        ],
        "className": "wrapper",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Ideas Board:",
            "className": "title",
          },
          "ref": null,
          "rendered": "Ideas Board:",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Sort by:",
              <select
                className="select"
                onChange={[Function]}
                value="Date"
              >
                <option
                  value="Date"
                >
                  Date
                </option>
                <option
                  value="Alphabet"
                >
                  Alphabet
                </option>
              </select>,
            ],
            "className": "label",
          },
          "ref": null,
          "rendered": Array [
            "Sort by:",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <option
                    value="Date"
                  >
                    Date
                  </option>,
                  <option
                    value="Alphabet"
                  >
                    Alphabet
                  </option>,
                ],
                "className": "select",
                "onChange": [Function],
                "value": "Date",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Date",
                    "value": "Date",
                  },
                  "ref": null,
                  "rendered": "Date",
                  "type": "option",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Alphabet",
                    "value": "Alphabet",
                  },
                  "ref": null,
                  "rendered": "Alphabet",
                  "type": "option",
                },
              ],
              "type": "select",
            },
          ],
          "type": "label",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`renders without crashing 2`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Header
    sortBy="Alphabet"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1
          className="title"
        >
          Ideas Board:
        </h1>,
        <label
          className="label"
        >
          Sort by:
          <select
            className="select"
            onChange={[Function]}
            value="Alphabet"
          >
            <option
              value="Date"
            >
              Date
            </option>
            <option
              value="Alphabet"
            >
              Alphabet
            </option>
          </select>
        </label>,
      ],
      "className": "wrapper",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Ideas Board:",
          "className": "title",
        },
        "ref": null,
        "rendered": "Ideas Board:",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Sort by:",
            <select
              className="select"
              onChange={[Function]}
              value="Alphabet"
            >
              <option
                value="Date"
              >
                Date
              </option>
              <option
                value="Alphabet"
              >
                Alphabet
              </option>
            </select>,
          ],
          "className": "label",
        },
        "ref": null,
        "rendered": Array [
          "Sort by:",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <option
                  value="Date"
                >
                  Date
                </option>,
                <option
                  value="Alphabet"
                >
                  Alphabet
                </option>,
              ],
              "className": "select",
              "onChange": [Function],
              "value": "Alphabet",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Date",
                  "value": "Date",
                },
                "ref": null,
                "rendered": "Date",
                "type": "option",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Alphabet",
                  "value": "Alphabet",
                },
                "ref": null,
                "rendered": "Alphabet",
                "type": "option",
              },
            ],
            "type": "select",
          },
        ],
        "type": "label",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1
            className="title"
          >
            Ideas Board:
          </h1>,
          <label
            className="label"
          >
            Sort by:
            <select
              className="select"
              onChange={[Function]}
              value="Alphabet"
            >
              <option
                value="Date"
              >
                Date
              </option>
              <option
                value="Alphabet"
              >
                Alphabet
              </option>
            </select>
          </label>,
        ],
        "className": "wrapper",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Ideas Board:",
            "className": "title",
          },
          "ref": null,
          "rendered": "Ideas Board:",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Sort by:",
              <select
                className="select"
                onChange={[Function]}
                value="Alphabet"
              >
                <option
                  value="Date"
                >
                  Date
                </option>
                <option
                  value="Alphabet"
                >
                  Alphabet
                </option>
              </select>,
            ],
            "className": "label",
          },
          "ref": null,
          "rendered": Array [
            "Sort by:",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <option
                    value="Date"
                  >
                    Date
                  </option>,
                  <option
                    value="Alphabet"
                  >
                    Alphabet
                  </option>,
                ],
                "className": "select",
                "onChange": [Function],
                "value": "Alphabet",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Date",
                    "value": "Date",
                  },
                  "ref": null,
                  "rendered": "Date",
                  "type": "option",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Alphabet",
                    "value": "Alphabet",
                  },
                  "ref": null,
                  "rendered": "Alphabet",
                  "type": "option",
                },
              ],
              "type": "select",
            },
          ],
          "type": "label",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
